[HLSL] Collect explicit resource binding information

hekota opened the pull request:
Scans each global variable declaration and its members and collects all required resource bindings in a new `SemaHLSL` data member `Bindings`.

New fields are added `HLSLResourceBindingAttr` for storing processed binding information so that it can be used by CodeGen (`Bindings` or any other Sema information is not accessible from CodeGen.)

Adjusts the existing register binding attribute handling and diagnostics to:
- do not create HLSLResourceBindingAttribute if it is not valid
- diagnose only the simple/local errors when a register binding attribute is parsed
- additional diagnostic of binding type mismatches is done later and uses the new `Bindings` data

Fixes #110719

hekota committed:
[HLSL] Collect explicit resource binding information (part 1)

- Do not create resource binding attribute if it is not valid
- Store basic resource binding information on HLSLResourceBindingAttr
- Move UDT type checking to to ActOnVariableDeclarator

Part 1 of #110719

hekota committed:
clang-format

bob80905 commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -985,44 +983,43 @@ SemaHLSL::TakeLocForHLSLAttribute(const HLSLAttributedResourceType *RT) {
-static bool ContainsResourceForRegisterType(Sema &S, const RecordType *RT,
-                                            RegisterType RegType) {
+// Walks though the user defined record type, finds resource class
+// that matches the RegisterBinding.Type and assigns it to
```
By RegisterBinding.Type you're referring to the type associated with RBA right? I think the comment could be clearer.

hekota replied:
Outdated.

bob80905 commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4588,6 +4588,35 @@ def HLSLResourceBinding: InheritableAttr {
+  let AdditionalMembers = [{
+      enum class RegisterType : unsigned { SRV, UAV, CBuffer, Sampler, C, I, Invalid };
+
+      const FieldDecl *ResourceField = nullptr;
```
Shouldn't this be made more abstract in the case of non-UDT resources? I would think the type would be Decl instead.

hekota replied:
I've removed the ResourceField.

bob80905 commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -1135,21 +1147,19 @@ static void ValidateMultipleRegisterAnnotations(Sema &S, Decl *TheDecl,
 
       RegisterType otherRegType = getRegisterType(attr->getSlot());
       if (RegisterTypesDetected[static_cast<int>(otherRegType)]) {
-        if (PreviousConflicts[TheDecl].count(otherRegType))
```
PreviousConflicts is no longer in use as a variable. You could remove it. But it seems unnecessary that you also handle duplicate RBAs in a separate part of the code. Wouldn't it be possible to check here if the decl is a UDT, and load in the relevant data into RBA (the resource field decl) if it is a UDT? 
If not, it might be useful to add a comment above the function stating: "This function validates multiple RBAs for non-udt resource decls, UDT resource decls are validated for duplicate RBAs in `ContainsResourceForRegisterType`"

hekota replied:
I have removed the unused variable. and the ResourceField is gone as well. I think the duplicate errors are reduced by simply not creating invalid binding attributes, but please let me know if you think otherwise. We should not be emitting duplicate errors for the same binding, but if there are multiple bindings with the same issue we should report them all.

bob80905 commented on clang/test/SemaHLSL/resource_binding_attr_error_udt.hlsl:
```diff
@@ -115,12 +114,14 @@ struct Eg13{
-// expected-warning@+4{{binding type 'u' only applies to types containing UAV resources}}
 // expected-warning@+3{{binding type 'u' only applies to types containing UAV resources}}
-// expected-warning@+2{{binding type 'u' only applies to types containing UAV resources}}
+// expected-error@+2{{binding type 'u' cannot be applied more than once}}
```
This extra diagnostic relates to my earlier comment, preventing multiple copies of the same diagnostic.

hekota replied:
Since there are 2 extra 'u' bindings we should be reporting 2 errors.

bob80905 commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -2228,3 +2242,62 @@ QualType SemaHLSL::getInoutParameterType(QualType Ty) {
+// Walks though existing explicit bindings, finds the actual resource class
+// decl the binding applies to and sets it to attr->ResourceField.
+// Additional processing of resource binding can be added here later on,
+// such as preparation for overapping resource detection or implicit binding.
```
overlapping

hekota replied:
Outdated

bob80905 commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -2228,3 +2242,62 @@ QualType SemaHLSL::getInoutParameterType(QualType Ty) {
+    if (!RBA)
+      continue;
+
+    // // Cbuffers and Tbuffers are HLSLBufferDecl types
```
double //

bob80905 commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -2228,3 +2242,62 @@ QualType SemaHLSL::getInoutParameterType(QualType Ty) {
+                                     ? ResourceClass::CBuffer
+                                     : ResourceClass::SRV) &&
+             "this should have been handled in DiagnoseLocalRegisterBinding");
+      // should we handle HLSLBufferDecl here?
```
I think so? It would definitely make sense if you were to change the additional member type in RBA to a Decl instead of a field decl, since HLSLBufferDecl hopefully inherits from Decl

bob80905 commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -2228,3 +2242,62 @@ QualType SemaHLSL::getInoutParameterType(QualType Ty) {
+      // FIXME: if array, calculate the binding size from the array dimensions
+      // (or unbounded for unsized array)
+      assert(RBA->getResourceField() == nullptr);
+      continue;
```
So, at some point in the future, you'd have something like `RBA->setResourceField(D)` here right? But it depends first on the FIXME?

hekota replied:
Outdated - ResourceField has been removed.

hekota committed:
Collect all resource binding requirements and analyze explicit bindings based on that
Also adds bindings size calculation and removed ResourceDecl field from HLSLResourceBindingAttr.

hekota committed:
removed unused variable, cleanup

hekota committed:
Merge branch 'main' into update-resource-binding-attr

damyanp commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4593,6 +4593,44 @@ def HLSLResourceBinding: InheritableAttr {
+      //-1 == unbounded
+      int Size;
+
+      void setBinding(RegisterType RT, unsigned SlotNum, unsigned SpaceNum, int Size = 0) {
```
`Size` doesn't seem to be used.  Is it still required?

hekota replied:
I'll remove the Size.

damyanp commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4593,6 +4593,44 @@ def HLSLResourceBinding: InheritableAttr {
+  let AdditionalMembers = [{
+      enum class RegisterType : unsigned { SRV, UAV, CBuffer, Sampler, C, I, Invalid };
+
+      RegisterType RegType;
+      unsigned SlotNumber;
+      unsigned SpaceNumber;
+      
+      // Size of the binding
+      // 0 == not set
+      //-1 == unbounded
+      int Size;
```
I can't find anywhere that unconditionally initializes these values?

damyanp replied:
It's suprising to me to see these member variables and the various accessors have the same visibility.  I'd expect something like the variables to be private and the accessors to be public.

damyanp commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4593,6 +4593,44 @@ def HLSLResourceBinding: InheritableAttr {
+        return SpaceNumber;
+      }
+      unsigned getSize() {
+        assert(Size == -1 || Size > 0 && "size not set");
```
> "size not set"

This implies that maybe Size is meant to be initialized to some value that's < -1 in order to be able to tell that it isn't set - yet I don't see that happening anywhere.

hekota replied:
I'll remove the size because at this point it is not being used or tested.

damyanp commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4593,6 +4593,44 @@ def HLSLResourceBinding: InheritableAttr {
+        Size = N;
+      }
+      bool isSizeUnbounded() {
+        return Size == -1;
```
Presumably this should have the same assert as on line 4623.  Or could be written as `return getSize() == -1;`

damyanp commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4593,6 +4593,44 @@ def HLSLResourceBinding: InheritableAttr {
+        SlotNumber = SlotNum;
+        SpaceNumber = SpaceNum;
+      }
+      RegisterType getRegisterType() {
```
All of these getters should be `const`.

damyanp commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -83,6 +81,49 @@ static RegisterType getRegisterType(StringRef Slot) {
+         "DeclBindingInfo already added");
+  if (DeclToBindingListIndex.find(VD) == DeclToBindingListIndex.end())
+    DeclToBindingListIndex[VD] = BindingsList.size();
+  return &BindingsList.emplace_back(DeclBindingInfo(VD, ResClass, Size));
```
```suggestion
  return &BindingsList.emplace_back(VD, ResClass, Size);
```

Or

```suggestion
  return &BindingsList.push_back(DeclBindingInfo(VD, ResClass, Size));
```

damyanp commented on clang/include/clang/Sema/SemaHLSL.h:
```diff
@@ -28,13 +28,62 @@ class AttributeCommonInfo;
+  const HLSLResourceBindingAttr *Attr;
+  BindingType BindType;
+
+  DeclBindingInfo(const VarDecl *Decl, ResourceClass ResClass, int Size = 0,
```
What does `Size==0` actually mean?  A zero sized array?

I'm wondering if `0` is actually a good default.  I notice that the only place this is called passes a value for Size anyway, so maybe it doesn't need default value.

hekota replied:
I'll remove the Size, it is not used or tested at this point.

damyanp commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -83,6 +81,49 @@ static RegisterType getRegisterType(StringRef Slot) {
+                                                      int Size) {
+  assert(getDeclBindingInfo(VD, ResClass) == nullptr &&
+         "DeclBindingInfo already added");
+  if (DeclToBindingListIndex.find(VD) == DeclToBindingListIndex.end())
+    DeclToBindingListIndex[VD] = BindingsList.size();
```
I think this'll work and only do a single lookup into the map:

```suggestion
  DeclToBindingListIndex.try_emplace(BindingsList.size());
```

hekota replied:
Love all these tricks with the DenseMap!

damyanp reviewed (commented):
I've looked through some of this and added some comments. I called out one place where there's some missing `const`s on members, but I think in general the code is not const correct and I didn't call out all the spots that I saw.

I have to admit I'm pretty confused about what the code is trying to do - the ratio of new code added to new tests / changes to the tests is a bit surprising. This means I can't look at the tests to try and understand what the code change is for, but it also points at maybe there being a testing gap.

bogner commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4593,6 +4593,44 @@ def HLSLResourceBinding: InheritableAttr {
   let Args = [StringArgument<"Slot">, StringArgument<"Space", 1>];
   let Documentation = [HLSLResourceBindingDocs];
+  let AdditionalMembers = [{
+      enum class RegisterType : unsigned { SRV, UAV, CBuffer, Sampler, C, I, Invalid };
```
Do we need to have `Invalid` in this enum? I don't particularly like the implication that we still need to check the validity of the binding once we're dealing with the attribute we've applied for codegen.

hekota replied:
This enum is just moved from SemaHLSL but yeah, there is no reason to keep the Invalid value here. It is only used when validating the register letter.

bogner commented on clang/include/clang/Sema/SemaHLSL.h:
```diff
@@ -102,6 +152,15 @@ class SemaHLSL : public SemaBase {
+  ResourceBindings Bindings;
+
+private:
+  void FindResourcesOnVarDecl(VarDecl *D);
+  void FindResourcesOnUserRecordDecl(const VarDecl *VD, const RecordType *RT,
+                                     int Size);
+  void ProcessExplicitBindingsOnDecl(VarDecl *D);
```
Better to name these startingWithLowerCase, which LLVM style generally prefers. It's a bit awkward in this area of the code because there's a lot of legacy naming, but if it isn't part of the Sema-interface parts of the class I think it makes sense to use the modern style.

hekota replied:
I was not sure what is the preferred way as the naming is sometimes all over the place.

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -83,6 +81,49 @@ static RegisterType getRegisterType(StringRef Slot) {
+    return ResourceClass::CBuffer;
+  case RegisterType::Sampler:
+    return ResourceClass::Sampler;
+  default:
+    llvm_unreachable("unexpected RegisterType value");
```
Let's explicitly spell out all of the enum values here instead of having a default case. This will play better with `-Wcovered-switch` if we add a new enum value at some point and forget to update this switch.

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -985,88 +1026,92 @@ SemaHLSL::TakeLocForHLSLAttribute(const HLSLAttributedResourceType *RT) {
+  return FindHandleTypeOnResource(VD->getType().getTypePtr());
+}
+
+// Walks though the global variable declaration, collects all resource binding
+// requirements and adds them to Bindings
+void SemaHLSL::FindResourcesOnUserRecordDecl(const VarDecl *VD,
+                                             const RecordType *RT, int Size) {
```
Might be better to call this "collectResourcesOnUserRecordDecl" - the verb "find" doesn't sound like it's going to mutate anything to me.

Also what is size here? It could probably use a mention in the comment.

hekota replied:
I'll remove the Size because it is not used or tested anywhere yet.

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -985,88 +1026,92 @@ SemaHLSL::TakeLocForHLSLAttribute(const HLSLAttributedResourceType *RT) {
-                                                 SourceLocation &ArgLoc,
-                                                 Decl *D, RegisterType RegType,
-                                                 bool SpecifiedSpace) {
+// return false if the register binding is not valid
+static bool DiagnoseLocalRegisterBinding(Sema &S, SourceLocation &ArgLoc,
+                                         Decl *D, RegisterType RegType,
+                                         bool SpecifiedSpace) {
```
If we're going to add a comment here it should really say what the function does, not just what the return value means.

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -2235,3 +2280,107 @@ QualType SemaHLSL::getInoutParameterType(QualType Ty) {
+  // Calculate size of array and unwrap
+  int Size = 1;
+  const Type *Ty = VD->getType()->getUnqualifiedDesugaredType();
+  if (Ty->isIncompleteArrayType())
+    Size = -1;
+  while (Ty->isConstantArrayType()) {
+    const ConstantArrayType *CAT = cast<ConstantArrayType>(Ty);
+    Size *= CAT->getSize().getSExtValue();
+    Ty = CAT->getElementType()->getUnqualifiedDesugaredType();
+  }
```
Is it possible to have an incomplete array type on one of the nested arrays here?

hekota replied:
No, an incomplete array can only be in first dimension: 

`RWBuffer<float> A[][10];` - ok
`RWBuffer<float> A[10][];` - DXC reports error

Clang currently does not allow either.

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -2235,3 +2280,107 @@ QualType SemaHLSL::getInoutParameterType(QualType Ty) {
+           "invalid or obsolete register type should never have an attribute "
+           "created");
+
+    // These were already diagnosed earlier
+    if (RT == RegisterType::C) {
+      if (Bindings.hasBindingInfoForDecl(VD))
+        SemaRef.Diag(VD->getLocation(),
+                     diag::warn_hlsl_user_defined_type_missing_member)
+            << static_cast<int>(RT);
+      continue;
+    }
```
If they were diagnosed earlier why are we emitting a diagnostic? Is this comment accurate?

hekota replied:
Nope, the comment needs to be removed.

hekota commented:
> I have to admit I'm pretty confused about what the code is trying to do - the ratio of new code added to new tests / changes to the tests is a bit surprising. This means I can't look at the tests to try and understand what the code change is for, but it also points at maybe there being a testing gap.

It is true that some of the information that is being gathered/calculated is not used anywhere yet, for example the size of the binding. It is likely going to be used on the future used when we implement implicit bindings or overlapping binding diagnostics, and maybe it will change to be something else. The question is - if it cannot be tested, should it be included now and complete https://github.com/llvm/llvm-project/issues/110719, or should it be added later as needed in the future binding-related work?

@bogner - do you have an opinion on this? @damyanp thinks it should be removed. I'm ok with replacing the size calculations with FIXME's.

bogner commented:
I think it's reasonable to limit this to implement things that are independently testable and leave the parts that aren't yet being used for some future change.

hekota committed:
Code review feedback

- remove size calculation and storage - it is currently not used or tested
- remove invalid register type
- set fields on HLSLResourceBindingAttr as private and accessors public, add const
- update function names
- update comments
- use more effective SmallVector and DenseMap methods

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into update-resource-binding-attr

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -985,88 +1034,94 @@ SemaHLSL::TakeLocForHLSLAttribute(const HLSLAttributedResourceType *RT) {
-      } else {
-        TypesToScan.emplace_back(FD->getType().getTypePtr());
-      }
+    // Field is a resource or array of resources
```
This comment isn't entirely accurate. At this point we could still have a struct or UDT that isn't a resource, right?

bogner commented on clang/include/clang/Sema/SemaHLSL.h:
```diff
@@ -28,13 +28,60 @@ class AttributeCommonInfo;
+  const HLSLResourceBindingAttr *Attr;
+  BindingType BindType;
+
+  DeclBindingInfo(const VarDecl *Decl, ResourceClass ResClass, int Size = 0,
```
Looks like you removed `Size` from most of this change but you missed the defaulted argument here.

hekota committed:
More cleanup - rename function to lowerCase and remove one overload, remove Size argument, remove comment

damyanp commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -58,31 +56,82 @@ static RegisterType getRegisterType(ResourceClass RC) {
+
+DeclBindingInfo *ResourceBindings::addDeclBindingInfo(const VarDecl *VD,
+                                                      ResourceClass ResClass) {
+  assert(getDeclBindingInfo(VD, ResClass) == nullptr &&
+         "DeclBindingInfo already added");
```
(This one worries me which is why I didn't click approve)

hekota replied:
I'll add a debug-only verification that existing bindings are stored sequentially and at the end of the BindingsList before adding a new one.

hekota commented on clang/include/clang/Sema/SemaHLSL.h:
```diff
@@ -28,13 +28,60 @@ class AttributeCommonInfo;
+public:
+  DeclBindingInfo *addDeclBindingInfo(const VarDecl *VD,
+                                      ResourceClass ResClass);
+  DeclBindingInfo *getDeclBindingInfo(const VarDecl *VD,
+                                      ResourceClass ResClass);
```
We'll need to leave `getDeclBindingInfo` const-less because we need to be able to modify the content of `DeclBindingInfo*` it returns, it cannot return `const DeclBindingInfo*`. I've tried to have it return `DeclBindingInfo const *` but that does not work.

damyanp replied:
Fair enough ... I tend to obsess over things like this and there are ways around it in this case - I'm not sure it'll actually improve the code in any objective way though!

hekota commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -58,31 +56,82 @@ static RegisterType getRegisterType(ResourceClass RC) {
-static RegisterType getRegisterType(StringRef Slot) {
+// Converts the first letter of string Slot to RegisterType.
+// Returns false if the letter does not correspond to a valid register type.
+static bool convertToRegisterType(StringRef Slot, RegisterType *RT) {
```
I've considered it, but since we `convertToRegisterType` does not create an error string or code that we should report I just kept it simple.

hekota replied:
Well, I tried to do it with `Expected<RegisterType>` and while does make the `convertToRegisterType` function smaller, checking of the results got more complicated because LLVM forces the error string/whatever to be checked. I'll leave it as is.

damyanp commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -985,88 +1034,85 @@ SemaHLSL::TakeLocForHLSLAttribute(const HLSLAttributedResourceType *RT) {
+      // Add a new DeclBindingInfo to Bindings if it does not already exist
+      ResourceClass RC = AttrResType->getAttrs().ResourceClass;
+      DeclBindingInfo *DBI = Bindings.getDeclBindingInfo(VD, RC);
+      if (!DBI)
```
When can this happen?  Does this mean we quietly ignore the case where the exact same binding is listed multiple times on a decl?

hekota replied:
This can happen when you have multiple resources of the same class in the same used record type. In that case we would update the size of the binding. This is how the code looked like before I removed the size calculations because they were not tested:

```
// Add a new DeclBindingInfo to Bindings. Update the binding size if
// a binding info already exists (there are multiple resources of same
// resource class in this user decl)
if (auto *DBI = Bindings.getDeclBindingInfo(VD, RC))
  DBI->Size += Size * ArraySize;
else
  Bindings.addDeclBindingInfo(VD, RC, Size);
```

damyanp replied:
At first glance something seems off about that, but without getting the actual array / size implementation in there it's hard to tell.

If there's no test coverage that exercises this case then it might be better to remove the `if` guard so that it doesn't accidentally mask some other problem we weren't expecting.

hekota replied:
The `if (!DBI)` case gets hit when parsing line 112 here where the user defined type `Eg12` has two resources classes of the same type: https://github.com/llvm/llvm-project/blob/5e9166e02ab65d42efba014f2adc59c42b097ddc/clang/test/SemaHLSL/resource_binding_attr_error_udt.hlsl#L112
(even though it is also testing duplicate register bindings).

hekota commented on clang/include/clang/Sema/SemaHLSL.h:
```diff
@@ -28,13 +28,60 @@ class AttributeCommonInfo;
+  // resource classes. They are all stored sequentially in this list.
+  // The DeclToBindingListIndex hashtable maps a declaration to the
+  // index of the first binding info in the list.
+  llvm::SmallVector<DeclBindingInfo> BindingsList;
+  llvm::DenseMap<const VarDecl *, unsigned> DeclToBindingListIndex;
```
Yeah. LLVM has `SparseMultiSet` class but it does not seem to be a good fit for our case:

*Fast multiset implementation for objects that can be identified by small unsigned keys.*
https://llvm.org/doxygen/classllvm_1_1SparseMultiSet.html

damyanp replied:
(I don't think it's worth changing, but I wondered if `std::multiset` or `std::unordered_multiset` would be reasonable candidates)

hekota committed:
cr feedback - add const, rearrange loop, add debug-only checks

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into update-resource-binding-attr

damyanp commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -110,6 +110,19 @@ DeclBindingInfo *ResourceBindings::addDeclBindingInfo(const VarDecl *VD,
                                                       ResourceClass ResClass) {
   assert(getDeclBindingInfo(VD, ResClass) == nullptr &&
          "DeclBindingInfo already added");
+#ifndef NDEBUG
+  // Verify that existing bindings for this decl are stored sequentially
+  // and at the end of the BindingsList
+  auto I = DeclToBindingListIndex.find(VD);
+  if (I != DeclToBindingListIndex.end()) {
+    for (unsigned Index = I->getSecond(); Index < BindingsList.size(); ++Index)
+      assert(BindingsList[Index].Decl == VD);
+  }
+#endif
```
I was thinking that `assert(!hasBindingInfoForDecl(VD) || BindingsList.back().Decl == VD);` would cover most bases?

hekota replied:
True

hekota replied:
Shoot! I have completed this PR before realizing I did not push this change up.
Here is a separate PR: https://github.com/llvm/llvm-project/pull/112661

damyanp reviewed (approved):
Some comments added, but I think we're probably good to go here.

hekota commented:
> If there's no test coverage that exercises this case then it might be better to remove the if guard so that it doesn't accidentally mask some other problem we weren't expecting.

The `if (!DBI)` case gets hit when parsing line 112 here where the user defined type `Eg12` has two resources classes of the same type: https://github.com/llvm/llvm-project/blob/5e9166e02ab65d42efba014f2adc59c42b097ddc/clang/test/SemaHLSL/resource_binding_attr_error_udt.hlsl#L112
(even though it is also testing duplicate register bindings).
